Make client/ directory (Calypso app) a Lerna package

jsnajdr opened the pull request:
Turns the Calypso client (and server) into proper Lerna monorepo package. The top-level `package.json` now contains only dev dependencies for the whole monorepo, and there is a new `client/package.json` that now has the runtime dependencies for Calypso client and server.

The main benefit is that we can now have a single unified `node_modules` tree, installed with one `npm ci` command, and with one global `package-lock.json`. Separate lockfiles and NPM installs in `apps/` were removed. The only part of the monorepo that continues to have its own NPM tree is `tests/e2e`.

The `server/` directory was moved to `client/server/` to make the Calypso app self-contained inside one directory. Many import paths in the `server/` sources needed to be updated and clarified.

The root `package.json` declares the app-like subpackages as `dependencies` to make them part of the `npm i` tree.

Some `devDependencies` were incorrectly declared as `dependencies`, so I moved them to the right place.

**Some areas that are not 100%**
- in `Dockerfile`, I had to disable the separate layer for `npm ci`. If I understand the logic correctly, it first copies a minimal version of the source tree that's needed to run the `npm install`, then installs the `node_modules`, and only then copies and builds the rest? @sirreal it would be nice if we could make it work again
- there are probably more `eslint-plugin-import` errors reported by ESLint now than needed. Can we improve on that?
- the `wp-desktop` is failing. 🙂 

**How to test:**
- test the Full Site Editing and WP.com Block Editor apps -- do they build and run as expected? Cc @Automattic/cylon as I personally have only limited knowledge about these apps.
- does Calypso really work 100%? The hardest part to make right were devdocs.

blowery commented on package.json:
```diff
@@ -34,189 +33,9 @@
-		"yargs": "13.3.0"
+		"@automattic/full-site-editing": "file:apps/full-site-editing",
+		"@automattic/wpcom-block-editor": "file:apps/wpcom-block-editor",
+		"wp-calypso": "file:client"
```
if we go this route we should probably include everything in apps/*?

jsnajdr replied:
Yes, and also all the packages in `packages/` that are not used by Calypso or something else in the repo transitively.

In other words, all packages in the monorepo should be direct or indirect dependencies of the root `package.json`. That ensures that `npm install` installs the whole package tree for the whole monorepo.

Gutenberg does it: https://github.com/WordPress/gutenberg/blob/master/package.json#L20-L66

Alternatively, `lerna bootstrap` does the npm installation process, but bails out to a simple `npm install` if it sees a local `file:` ref in `dependencies` in the root `package.json`. We could also go that way, but leaving the npm tree solely to the npm client is preferrable.

blowery replied:
> Alternatively, lerna bootstrap does the npm installation process, but bails out to a simple npm install if it sees a local file: ref in dependencies in the root package.json. We could also go that way, but leaving the npm tree solely to the npm client is preferrable.

Yeah, if we want to use lerna bootstrap, we'd also want to jump back to semver ranges instead of file refs. Once you add one file ref bootstrap stops making sense. If we really want a mix, something like yarn workspaces or pnpm workspaces offer a lot more flexibility.

blowery replied:
The way I've been thinking about it: lerna provides the "I want to run a command against the package graph" functionality and something else (npm, yarn, pnpm) provides the install and graph linkage. `lerna bootstrap` is to be avoided as it's just a hacky way of doing what one of the package managers would rather do itself.

jsnajdr replied:
Totally agreed that `lerna bootstrap` is to be avoided. Even the Lerna maintainer told us that it's a buggy and hard-to-maintain hack that tries to do the package manager's job (poorly).

simison commented:
cc @tyxla this PR might be relevant movement to follow regarding Jetpack Cloud living in `/apps`.

jsnajdr committed:
Move the validate-config-keys script (used in pre-commit hook) to bin/

jsnajdr committed:
Move webpack utility script for extensions aliases to client/webpack/

That's also a great future location for all custom webpack-related modules. Better than
the currently used `server/bundler`.

jsnajdr committed:
Remove lodash dependency from webpack configs

jsnajdr committed:
Server analytics: improve imports from client dir

jsnajdr committed:
Move dev dependencies to top-level devDependencies

jsnajdr committed:
calypso-build: call npx with no-install, make enzyme a peer dependency

jsnajdr committed:
FSE: call wp-scripts with npx so that the bin location doesn't matter

jsnajdr committed:
Dockerfile: disable the separate npm ci layer

jsnajdr committed:
FSE: don't npm ci before FSE build. There's only one global npm ci

jsnajdr committed:
Remove separate apps/ lockfiles. There is only one global one

jsnajdr committed:
Simplify imports from client/directory (use the short form)

jsnajdr committed:
Move server/ to client/server/, update paths

jsnajdr committed:
Add client/ package to lerna config

jsnajdr committed:
Create client/package.json, move non-dev dependencies there

jsnajdr committed:
Update the (unified) lockfile

jsnajdr committed:
Fixup paths for ICFY stats

jsnajdr commented:
This seems to work pretty well at this moment and I think we are ready for review and testing 😊

ockham commented:
Would we consider (possibly as a follow-up) `git mv client/ apps/calypso/`?

noahtallen commented:
@jsnajdr are there particular changes which could affect FSE here? (i.e. any idea what I should look out for when testing?) I see that the package-lock files for the individual apps are removed, which is nice :)

jsnajdr commented:
> are there particular changes which could affect FSE here? (i.e. any idea what I should look out for when testing?)

There should be no user-observable change in the app behavior. It's only the build and test workflow that's been updated. Please look for bugs and breakages in the development workflow :+1:

> I see that the package-lock files for the individual apps are removed, which is nice :)

Yes, that solves a lot of intractable problems with running `npm install` on the repo. See the discussion in https://github.com/Automattic/wp-calypso/pull/38001#issuecomment-561617033 to see what motivated to get this task finally done 🙂

jsnajdr commented:
> Would we consider (possibly as a follow-up) `git mv client/ apps/calypso/`?

Yes! But that's a PR that affects thousands of files and possibly invalidates all open PRs. (I'm not sure if Git can detect a rename and a modification of a file in different branches and automatically merge everything together).

One property of this PR that I'm very happy about is that it's relatively small -- 70 modified files, most of them simple renames from `client/` to `client/server/`.

After the devdocs cleanups in #38784 and #38785 are merged, it will become even smaller 🥳

mmtr commented:
Verified the `wpcom-block-editor` app still builds correctly both locally (`npx lerna run build --scope='@automattic/wpcom-block-editor' -- -- --output-path=./dist`) and on [CircleCI](https://circleci.com/gh/Automattic/wp-calypso/562290).

I tested both builds on my sandbox and confirmed that all their features work as usual.

jsnajdr committed:
Fixup path in devdocs/selectors

jsnajdr committed:
Devdocs: fixup import and externalizing of the search-index module

The request string and the externals declaration must exactly match. That makes
sure that `search-index.js` is not bundled into server bundle, but continues to be
read with `require( 'server/bundler/search-index')`.

Also removes legacy externals like `hot-reloader` or `bundler/assets` (the assets file
is read with `fs.readFile`, not with Node require)

jsnajdr committed:
ESLint: fixup the configuration of the import/no-extraneous-modules

Don't specify a `packageDir` and make it look at the closest `package.json` by default.

Also, add `lodash` as a top-level dev dependency, as some build-only files (e.g., `.eslintrc`)
use it.

jsnajdr committed:
Update the npm-package-json-lint configuration and runner

Allow `devDependencies` only in the root `package.json` and not anywhere else.

jsnajdr commented:
- `wp-desktop` build works after the latest updates, and will also require https://github.com/Automattic/wp-desktop/pull/739 to be merged
- the `import/no-extraneous-dependencies` ESLint rule now also works correctly and doesn't report non-issues

jsnajdr commented:
Rebased onto the latest master, with #38660 (CoreJS 3 upgrade) and #38531 (`node_modules` transpiling in `calypso-build`) merged in.

blowery reviewed (approved):
This looks good to me. Worked in all the common cases I typically do.

noahtallen reviewed (approved):
Looks good from the FSE side of things! I tested the following:

1. Running our wpcom sync script using the CircleCi build works as expected. (D37756-code is the diff that it created, which looks correct to me.)
2. Running the dev command locally works as expected with everything built correctly.

I do have some questions about monorepo type things:
- Should we always run `npm i` from wp-calypso root?
- Should all dependencies (both dev and prod) be added to the root `package.json` instead of the app-level package.json?

jsnajdr commented:
> Should we always run `npm i` from `wp-calypso` root?

Yes. There is now only one NPM tree, are package subdirectories are linked to the root `package.json` with `file:` references.

> Should all dependencies (both dev and prod) be added to the root package.json instead of the app-level package.json?

dev depedencies should be always in the root `package.json`. They are common for the whole project. Prod dependencies should be always in the leaf `package.json`.

I'll post a long form P2 with a detailed rationale and bit of theory behind these answers 🙂

ockham commented:
:wave: @jsnajdr I noticed that this PR moved the root Webpack configs to `client/`, but not the Babel and PostCSS ones. Shouldn't those also be moved there? They're only for Calypso-the-app, aren't they?

jsnajdr commented:
> I noticed that this PR moved the root Webpack configs to `client/`, but not the Babel and PostCSS ones. Shouldn't those also be moved there?

That's a good question! When I tried to move `babel.config.js` to `client/`, I couldn't make Jest work. The `babel-jest` plugin is a zero-config thing and it couldn't find the Babel config in `client/`.

The `babel.config.js` file is supposed to live in the project root, or more precisely, in the `cwd` of the Node process that runs the build scripts. And Calypso build scripts still live in the root `package.json` and are ran from the repo's root directory.

A possible solution is to never allow the `babel.config.js` lookup algorithm to run and always specify the config explicitly as JS object. We already do that in the webpack configs, but there are places like `babel-jest` that need to find the file.

I don't know if the PostCSS file can be safely moved. I forgot about it when working on the patch, as the CSS build was never broken because of it 🙂
